### ISSUE TITLE
Assessment and Keyable class

### ISSUE THREAD

**Issue by iztanpy** [nus-cs2030/2021-s2#70]
# Summary

I have a problem with the output of the get returning me a Keyable instead of an assessment for one of the module test cases. However, when I run getClass I see that it is an Assessment class.

## Description:

Clarifications or details with respect to the issue you are facing

## Code Snippets

Attach code snippets (if any)

## Screenshots

<img width="1440" alt="Screenshot 2021-03-13 at 10 06 56 PM" src="https://user-images.githubusercontent.com/61683649/111032755-bdb1f400-8448-11eb-85d2-a2387c27367a.png">

**Comment by Keithlimhs** [comment]
Does your get method in KeyableMap returns a KeyableMap<V> ?

**Comment by iztanpy** [comment]
My get method in KeyableMap returns a V which is the generic type

**Comment by Keithlimhs** [comment]
Your Assessment class implements Keyable interface right?

**Comment by iztanpy** [comment]
yup it does and so does my V generic

**Comment by Keithlimhs** [comment]
I think it may be your put method in KeyableMap class. Did you put your return type as KeyableMap instead of KeyableMap<V>?

**Comment by iztanpy** [comment]
hmm I don't think it is though I returned the correct type

**Comment by xnajasho** [comment]
Your module class should extend a KeyableMap of type V which implements Keyable

**Comment by Keithlimhs** [comment]
oh so your return type for put in KeyableMap is KeyableMap< V >? Cos i changed my return type from KeyableMap< V > to KeyableMap and i got the same error

**Comment by iztanpy** [comment]
yup for me my return types KeyableMap< v> too, not sure why got this problem

**Comment by iztanpy** [comment]
I think from what I understand V extends a Keyable, so during compilation time, it sets it to a Keyable instead of an Assessment, which causes the problem, but im not sure how to fix it

**Comment by clementyee303** [comment]
If your get() method in your KeyableMap<V> class is returning a KeyableMap<V> / KeyableMap, you should rethink your implementation. Recall that KeyableMap models an entity that contains a map

**Comment by Keithlimhs** [comment]
> I think from what I understand V extends a Keyable, so during compilation time, it sets it to a Keyable instead of an Assessment, which causes the problem, but im not sure how to fix it

@iztanpy yea but it shouldnt cause a compilation error since your Assessment implements Keyable as well. May i know how your constructor for KeyableMap and module looks like?

**Comment by iztanpy** [comment]
public KeyableMap(String name) {
        this.name = name;
        this.map = new HashMap<String, V>();
    }

public class Module extends KeyableMap implements Keyable{
    @SuppressWarnings("unchecked")
    public Module(String name) {
        super(name, new HashMap<String, Assessment>());
    }

**Comment by miao-tianjia** [comment]
your put method should return a keyablemap<V> and your get method should return a V. then assessment, module, and student should all implement Keyable interface. lastly, Module should extend KeyableMap<Assessment> and Student should extend KeyableMap<Module>. as I cannot see your actual code I'm not very sure where your code goes wrong...

**Comment by miao-tianjia** [comment]
> your put method should return a keyablemap and your get method should return a V. then assessment, module, and student should all implement Keyable interface. lastly, Module should extend KeyableMap and Student should extend KeyableMap. as I cannot see your actual code I'm not very sure where your code goes wrong...

Sorry, the put method should return a KeyableMap<V>, I got a typo there.

**Comment by iztanpy** [comment]
protected KeyableMap(String name, HashMap<String, V> map){
 14         this.name = name;
 15         this.map = map;
 16     }

  public V get(String search) {
 23         return (V)map.get(search);
 24     }


@SuppressWarnings("unchecked")
 45     public KeyableMap<V> put(V item) {
 46         map.put(item.getKey(),item);
 47         return new KeyableMap(this.name, this.map);
 48     }

**Comment by Keithlimhs** [comment]
> public KeyableMap(String name) {
> this.name = name;
> this.map = new HashMap<String, V>();
> }
> 
> public class Module extends KeyableMap implements Keyable{
> @SuppressWarnings("unchecked")
> public Module(String name) {
> super(name, new HashMap<String, Assessment>());
> }

Your constructor in KeyableMap here is alright. When your module extends KeyableMap you need to include what is the V. So it would be Module extends KeyableMap<Assessment> implements keyable

**Comment by Keithlimhs** [comment]
> protected KeyableMap(String name, HashMap<String, V> map){
> 14 this.name = name;
> 15 this.map = map;
> 16 }
> 
> public V get(String search) {
> 23 return (V)map.get(search);
> 24 }
> 
> @SuppressWarnings("unchecked")
> 45 public KeyableMap put(V item) {
> 46 map.put(item.getKey(),item);
> 47 return new KeyableMap(this.name, this.map);
> 48 }

For your put method here you do not need to return a new KeyableMap< V > as the HashMap is mutable. So you can just return this

**Comment by iztanpy** [comment]
> "

Yup this was the problem thank you so much

**Comment by malcolmsng** [comment]
It should be like this:

![image](https://user-images.githubusercontent.com/75665163/111893970-af179e00-8a41-11eb-933b-a880a77f194c.png)

**Comment by JWulaXia** [comment]
my idea is that under class Module, put should take in Assessment type and return class Module, and get method should return Assessment type, if you encounter type conversion error, you should go through you class structure all again, check which returning type is wrong.
